### PR TITLE
PLU-125: [EXCEL] Update table row

### PR DESCRIPTION
**Pull request by ogp-weeloong**
## Problem
Users need a way to update a row in their excel table.

## Solution
Implement an action to do this. Similar to our vault workspace "update table data" action, this allows users to update the 1st row whose lookup column matches the lookup value.

The initial lookup is exactly the same as `getTableRow` action, so we re-use its implementation.

NOTE: This also removes `tableRowNumber` to prevent user confusion. It's not being used anywhere, so safe to remove. I will add it back in if users ask for it.

## Tests
- Check that row update works (i.e. it updates the 1st / topmost row whose lookup column matches the lookup value).
- Check that it returns the updated row data, including output from formulas.
- Check that it clears cells if the update value is blank.
- Check that "Found Rows" is false if a viable row is not found.
- Check that it indeed fetches `MAX_ROW`s rows.
- Check that it errors out if the table is too large.
- Check that it errors out if the lookup column is missing.

### CODE REVIEW

File: packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+      lookupValue,
+      lookupColumn,
+    })
+
+    if (!findRowResults) {
+      $.setActionItem({
+        raw: {
+          updatedRow: false,
+        } satisfies DataOut,
+      })
```
Comment (pregnantboy) [inline packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts:114]: need to return here

File: packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    const { tableRowIndex, sheetRowNumber, columns } = findRowResults
+
+    //
+    // Perform update...
+    //
+
+    // Return updated row in case it has formulas.
+    const updateRowValuesParseResult = updateRowValuesResponseSchema.safeParse(
+      (
+        await session.request(
```
Comment (pregnantboy) [inline packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts]: nit: dont like to nest function calls, makes it hard to read

**Review by pregnantboy — DISMISSED** [review]
somehow need to specify that it only updates the first row that matches the lookup

File: packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts (hunk 1 above)
Comment (ogp-weeloong) [reply packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts:114]: fml good catch. thanks!

File: packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  parametersSchema,
+  updateRowValuesResponseSchema,
+} from './schemas'
+
+type DataOut = Required<z.infer<typeof dataOutSchema>>
+
+const action: IRawAction = {
+  name: 'Update table row',
+  key: 'updateTableRow',
+  description: 'Updates a single row of data in your Excel table',
```
Comment (ogp-weeloong) [inline packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts:27]: @pregnantboy what do you think of `Updates the first matching row of data in your Excel table`.

Comment (ogp-weeloong) [reply packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts:27]: Have updated copy in getTableRow action instead. Pull and lmk what you think

File: packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts (hunk 2 above)
Comment (ogp-weeloong) [reply packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts]: changed

File: packages/backend/src/apps/m365-excel/actions/update-table-row/schemas.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        columnName: z
+          .string()
+          .min(1, { message: 'Please select a column to update.' }),
+        // We allow empty strings to support optional form fields.
+        value: z.string(),
+      }),
+    )
+    .min(1, { message: 'Please input at least 1 column to update.' })
+    .superRefine((columnsToUpdate, context) => {
+      // Shoud not have repeated columns
```
Comment (m0nggh) [inline packages/backend/src/apps/m365-excel/actions/update-table-row/schemas.ts:20]: wah nice catch!

File: packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        } satisfies DataOut,
+      })
+
+      return
+    }
+
+    const { tableRowIndex, sheetRowNumber, columns } = findRowResults
+
+    //
+    // Perform update...
```
Comment (m0nggh) [inline packages/backend/src/apps/m365-excel/actions/update-table-row/index.ts:122]: noob question: can i ask what's this for ah?

**Review by m0nggh — COMMENTED** [review]
tested:
- update first row whose lookup column matches the lookup value
- update works even if same update value is given
- negative case works: no found row
- errors work e.g. same update column used, update column missing after being deleted from the table

good job, will let ian approve the description part and other stuff if needed

**Review by m0nggh — APPROVED** [review]
lgtm other than the wording part!

**Comment by ogp-weeloong** [comment]
@pregnantboy dismissed your review as I've addressed your comments and checked by malcolm. Should not need you to look again since you busy. Feel free to look / comment later.

**Pull request merged**